File: api/handlers/distribution.go
/**
 * 模块说明：数字丝路分发接口控制器。
 * 业务场景：前端需要管理社媒授权目标，并把生成的图文/视频提交到多平台异步分发。
 * 核心职责：按设备身份隔离目标与任务，绑定请求参数，并把业务错误转成前端可理解的响应。
 */
package handlers

import (
	"errors"
	"strconv"
	"strings"

	middlewares2 "github.com/drama-generator/backend/api/middlewares"
	services2 "github.com/drama-generator/backend/application/services"
	"github.com/drama-generator/backend/pkg/logger"
	"github.com/drama-generator/backend/pkg/response"
	"github.com/gin-gonic/gin"
	"gorm.io/gorm"
)

type DistributionHandler struct {
	service *services2.DistributionService
	log     *logger.Logger
}

func NewDistributionHandler(service *services2.DistributionService, log *logger.Logger) *DistributionHandler {
	return &DistributionHandler{
		service: service,
		log:     log,
	}
}

/**
 * 功能：查询当前设备的分发账号和目标。
 * 参数：c 为 Gin 请求上下文，设备 ID 从中间件读取。
 * 返回：targets 包含 Upload-Post profile、Pinterest board、Reddit subreddit 和 Discord webhook。
 */
func (h *DistributionHandler) ListTargets(c *gin.Context) {
	deviceID := middlewares2.GetDeviceID(c)

	targets, err := h.service.ListTargets(deviceID)
	if err != nil {
		h.respondError(c, err)
		return
	}

	response.Success(c, gin.H{"targets": targets})
}

/**
 * 功能：创建或确认 Upload-Post profile。
 * 参数：c 为 Gin 请求上下文。
 * 返回：profile；Pinterest/Reddit 授权链接依赖它建立外部账号归属。
 */
func (h *DistributionHandler) EnsureUploadPostProfile(c *gin.Context) {
	deviceID := middlewares2.GetDeviceID(c)

	profile, err := h.service.EnsureUploadPostProfile(c.Request.Context(), deviceID)
	if err != nil {
		h.respondError(c, err)
		return
	}

	response.Success(c, gin.H{"profile": profile})
}

/**
 * 功能：生成第三方平台授权连接链接。
 * 参数：c 为 Gin 请求上下文。
 * 返回：profile 与 access_url，前端打开 access_url 完成 Pinterest/Reddit 授权。
 */
func (h *DistributionHandler) GenerateUploadPostConnectLink(c *gin.Context) {
	deviceID := middlewares2.GetDeviceID(c)

	profile, accessURL, err := h.service.GenerateUploadPostConnectLink(c.Request.Context(), deviceID)
	if err != nil {
		h.respondError(c, err)
		return
	}

	response.Success(c, gin.H{
		"profile":    profile,
		"access_url": accessURL,
	})
}

func (h *DistributionHandler) SyncUploadPostProfile(c *gin.Context) {
	deviceID := middlewares2.GetDeviceID(c)

	profile, err := h.service.SyncUploadPostProfile(c.Request.Context(), deviceID)
	if err != nil {
		h.respondError(c, err)
		return
	}

	response.Success(c, gin.H{"profile": profile})
}

func (h *DistributionHandler) ListPinterestBoards(c *gin.Context) {
	deviceID := middlewares2.GetDeviceID(c)

	boards, err := h.service.ListPinterestBoards(c.Request.Context(), deviceID)
	if err != nil {
		h.respondError(c, err)
		return
	}

	response.Success(c, gin.H{"boards": boards})
}

func (h *DistributionHandler) SetDefaultTarget(c *gin.Context) {
	deviceID := middlewares2.GetDeviceID(c)
	targetID, err := strconv.ParseUint(c.Param("id"), 10, 32)
	if err != nil {
		response.BadRequest(c, "无效的 target ID")
		return
	}

	target, err := h.service.SetDefaultTarget(deviceID, uint(targetID))
	if err != nil {
		h.respondError(c, err)
		return
	}

	response.Success(c, gin.H{"target": target})
}

func (h *DistributionHandler) SaveRedditDefaultTarget(c *gin.Context) {
	deviceID := middlewares2.GetDeviceID(c)

	var req services2.UpsertRedditTargetRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		response.BadRequest(c, "无效的 subreddit 配置")
		return
	}

	target, err := h.service.SaveRedditDefaultTarget(deviceID, req)
	if err != nil {
		h.respondError(c, err)
		return
	}

	response.Success(c, gin.H{"target": target})
}

func (h *DistributionHandler) UpsertDiscordTarget(c *gin.Context) {
	deviceID := middlewares2.GetDeviceID(c)

	var req services2.UpsertDiscordTargetRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		response.BadRequest(c, "无效的 Discord 配置")
		return
	}

	target, err := h.service.UpsertDiscordTarget(c.Request.Context(), deviceID, req)
	if err != nil {
		h.respondError(c, err)
		return
	}

	response.Success(c, gin.H{"target": target})
}

func (h *DistributionHandler) DeleteTarget(c *gin.Context) {
	deviceID := middlewares2.GetDeviceID(c)
	targetID, err := strconv.ParseUint(c.Param("id"), 10, 32)
	if err != nil {
		response.BadRequest(c, "无效的 target ID")
		return
	}

	if err := h.service.DeleteTarget(deviceID, uint(targetID)); err != nil {
		h.respondError(c, err)
		return
	}

	response.Success(c, gin.H{"message": "target 已删除"})
}

/**
 * 功能：创建一次异步分发任务。
 * 参数：c 为 Gin 请求上下文，Body 包含内容、媒体、平台、目标和发布时间。
 * 返回：job，包含每个平台结果的初始状态。
 */
func (h *DistributionHandler) CreateDistribution(c *gin.Context) {
	deviceID := middlewares2.GetDeviceID(c)

	var req services2.CreateDistributionRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		response.BadRequest(c, "无效的分发请求")
		return
	}

	job, err := h.service.CreateDistribution(c.Request.Context(), deviceID, &req)
	if err != nil {
		h.respondError(c, err)
		return
	}

	response.Success(c, gin.H{"job": job})
}

func (h *DistributionHandler) ListDistributionJobs(c *gin.Context) {
	deviceID := middlewares2.GetDeviceID(c)
	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	pageSize, _ := strconv.Atoi(c.DefaultQuery("page_size", "20"))

	jobs, total, err := h.service.ListDistributionJobs(deviceID, page, pageSize)
	if err != nil {
		h.respondError(c, err)
		return
	}

	response.Success(c, gin.H{
		"jobs":      jobs,
		"total":     total,
		"page":      page,
		"page_size": pageSize,
	})
}

func (h *DistributionHandler) GetDistributionJob(c *gin.Context) {
	deviceID := middlewares2.GetDeviceID(c)
	jobID, err := strconv.ParseUint(c.Param("id"), 10, 32)
	if err != nil {
		response.BadRequest(c, "无效的 job ID")
		return
	}

	job, err := h.service.GetDistributionJob(deviceID, uint(jobID))
	if err != nil {
		h.respondError(c, err)
		return
	}

	response.Success(c, gin.H{"job": job})
}

func (h *DistributionHandler) RetryDistribution(c *gin.Context) {
	deviceID := middlewares2.GetDeviceID(c)
	jobID, err := strconv.ParseUint(c.Param("id"), 10, 32)
	if err != nil {
		response.BadRequest(c, "无效的 job ID")
		return
	}

	job, err := h.service.RetryDistribution(c.Request.Context(), deviceID, uint(jobID))
	if err != nil {
		h.respondError(c, err)
		return
	}

	response.Success(c, gin.H{"job": job})
}

/**
 * 功能：统一分发业务错误响应。
 * 参数：c 为 Gin 请求上下文；err 为服务层错误。
 * 返回：无返回值；可修复的配置/参数问题返回 400，未知错误返回 500。
 */
func (h *DistributionHandler) respondError(c *gin.Context, err error) {
	switch {
	case errors.Is(err, gorm.ErrRecordNotFound):
		response.NotFound(c, "资源不存在")
	case isDistributionBadRequest(err):
		response.BadRequest(c, err.Error())
	default:
		h.log.Errorw("Distribution request failed", "error", err)
		response.InternalError(c, err.Error())
	}
}

func isDistributionBadRequest(err error) bool {
	if err == nil {
		return false
	}

	message := err.Error()
	for _, marker := range []string{
		"不支持",
		"不能为空",
		"无效",
		"请先",
		"需要",
		"暂不支持",
	} {
		if strings.Contains(message, marker) {
			return true
		}
	}

	return false
}

File: web/src/api/distribution.ts
/**
 * 模块说明：数字丝路多平台分发接口封装。
 * 业务场景：生成的图文/视频素材需要发布到 Pinterest、Reddit、Discord 等外部渠道。
 * 核心职责：管理分发账号、目标、任务创建、任务查询和失败重试，隐藏后端返回包装结构。
 */
import request from '../utils/request'
import type {
  CreateDistributionRequest,
  DistributionJob,
  DistributionTarget,
  DistributionTargetsView,
  UploadPostProfile
} from '@/types/distribution'

export interface UploadPostConnectLinkResponse {
  profile: UploadPostProfile
  access_url: string
}

export interface UpsertDiscordTargetRequest {
  webhookUrl: string
  name?: string
  isDefault?: boolean
}

export interface UpsertRedditTargetRequest {
  subreddit: string
  flairId?: string
}

export const distributionAPI = {
  /**
   * 功能：获取当前设备身份下的分发账号和目标。
   * 参数：无。
   * 返回：Upload-Post profile 与各平台目标列表，用于判断能否提交分发。
   */
  async listTargets(): Promise<DistributionTargetsView> {
    const response = await request.get<{ targets: DistributionTargetsView }>('/distributions/targets')
    return response.targets
  },

  /**
   * 功能：确保当前设备有 Upload-Post profile。
   * 参数：无。
   * 返回：UploadPostProfile，后续 Pinterest/Reddit 授权链接都依赖该 profile。
   */
  async ensureUploadPostProfile(): Promise<UploadPostProfile> {
    const response = await request.post<{ profile: UploadPostProfile }>('/distributions/upload-post/profile/ensure')
    return response.profile
  },

  /**
   * 功能：同步 Upload-Post 授权状态。
   * 参数：无。
   * 返回：最新 profile，包含已连接平台和最近同步时间。
   */
  async syncUploadPostProfile(): Promise<UploadPostProfile> {
    const response = await request.post<{ profile: UploadPostProfile }>('/distributions/upload-post/sync')
    return response.profile
  },

  /**
   * 功能：生成外部平台授权连接页。
   * 参数：无。
   * 返回：profile 与 access_url，前端打开 access_url 让用户完成 Pinterest/Reddit 授权。
   */
  async generateUploadPostConnectLink(): Promise<UploadPostConnectLinkResponse> {
    return request.post<UploadPostConnectLinkResponse>('/distributions/upload-post/connect-link')
  },

  async listPinterestBoards(): Promise<DistributionTarget[]> {
    const response = await request.get<{ boards: DistributionTarget[] }>('/distributions/pinterest/boards')
    return response.boards || []
  },

  async setDefaultTarget(targetId: number): Promise<DistributionTarget> {
    const response = await request.put<{ target: DistributionTarget }>(`/distributions/targets/${targetId}/default`)
    return response.target
  },

  async saveRedditDefaultTarget(payload: UpsertRedditTargetRequest): Promise<DistributionTarget> {
    const response = await request.put<{ target: DistributionTarget }>('/distributions/targets/reddit/default', payload)
    return response.target
  },

  async upsertDiscordTarget(payload: UpsertDiscordTargetRequest): Promise<DistributionTarget> {
    const response = await request.post<{ target: DistributionTarget }>('/distributions/targets/discord', payload)
    return response.target
  },

  async deleteTarget(targetId: number): Promise<void> {
    await request.delete(`/distributions/targets/${targetId}`)
  },

  async listJobs(params?: { page?: number; page_size?: number }): Promise<{ jobs: DistributionJob[]; total: number }> {
    const response = await request.get<{ jobs: DistributionJob[]; total: number }>('/distributions', { params })
    return {
      jobs: response.jobs || [],
      total: response.total || 0
    }
  },

  /**
   * 功能：创建一次多平台分发任务。
   * 参数：payload 包含内容类型、来源、媒体地址、目标平台、平台参数和发布模式。
   * 返回：DistributionJob，包含每个平台的执行结果初始状态。
   */
  async createDistribution(payload: CreateDistributionRequest): Promise<DistributionJob> {
    const response = await request.post<{ job: DistributionJob }>('/distributions', payload)
    return response.job
  },

  /**
   * 功能：查询单个分发任务最新状态。
   * 参数：jobId 为后端分发任务 ID。
   * 返回：DistributionJob，前端轮询它来展示各平台成功、失败或待处理状态。
   */
  async getJob(jobId: number): Promise<DistributionJob> {
    const response = await request.get<{ job: DistributionJob }>(`/distributions/${jobId}`)
    return response.job
  },

  async retryJob(jobId: number): Promise<DistributionJob> {
    const response = await request.post<{ job: DistributionJob }>(`/distributions/${jobId}/retry`)
    return response.job
  }
}
